File: go_snippets/bipartite_matching.go
package go_snippets

type BipartiteMatching struct {
	edges [][]int
	match []int
	used  []bool
}

func (b *BipartiteMatching) init(n int) {
	b.edges = make([][]int, n)
	b.match = make([]int, n)

	for i := range b.match {
		b.match[i] = -1
	}
}

func (b *BipartiteMatching) add(from, to int) {
	b.edges[from] = append(b.edges[from], to)
	b.edges[to] = append(b.edges[to], from)
}

func (b *BipartiteMatching) dfs(v int) bool {
	b.used[v] = true

	for _, u := range b.edges[v] {
		w := b.match[u]
		if w == -1 || !b.used[w] && b.dfs(w) {
			b.match[u] = v
			b.match[v] = u
			return true
		}
	}

	return false
}

func (b *BipartiteMatching) exec() int {
	res := 0

	for v := 0; v < len(b.edges); v++ {
		if b.match[v] == -1 {
			b.used = make([]bool, len(b.edges))

			if b.dfs(v) {
				res++
			}
		}
	}

	return res
}

File: snippets/cpp/bipartite_matching.cpp
#include "algorithm"
#include "iostream"
#include "map"
#include "vector"
#define rep(i, to) for (ll i = 0; i < (to); ++i)
using namespace std;
typedef long long ll;

struct BipartiteMatching {
  vector<vector<ll>> edges;
  vector<ll> match;
  vector<bool> used;

  BipartiteMatching(ll n)
      : edges(vector<vector<ll>>(n)), match(vector<ll>(n, -1)) {}

  void add(ll from, ll to) {
    edges[from].push_back(to);
    edges[to].push_back(from);
  }

  bool dfs(ll v) {
    used[v] = true;

    for (auto u : edges[v]) {
      ll w = match[u];
      if (w == -1 || !used[w] && dfs(w)) {
        match[u] = v;
        match[v] = u;
        return true;
      }
    }

    return false;
  }

  ll call() {
    ll res = 0;

    rep(v, edges.size()) {
      if (match[v] == -1) {
        used = vector<bool>(edges.size());

        if (dfs(v)) {
          res++;
        }
      }
    }

    return res;
  }
};
